libsql-sqlite3/src/vectorIndex.c

```diff
@@ -0,0 +1,462 @@
+
+int vectorIdxKeyGet(Table *pTable, VectorIdxKey *pKey, const char **pzErrMsg) {
+  int i;
+  Index *pPk;
+  // we actually need to change strategy here and use PK if it's available and fallback to ROWID only if there is no other choice
+  // will change this later as it must be done carefully in order to not brake behaviour of existing indices
+  if( !HasRowid(pTable) ){
+    pPk = sqlite3PrimaryKeyIndex(pTable);
+    if( pPk->nKeyCol > VECTOR_INDEX_MAX_KEY_COLUMNS ){
+      *pzErrMsg = "exceeded limit for composite columns in primary key index";
+      return -1;
+    }
+    pKey->nKeyColumns = pPk->nKeyCol;
+    for(i = 0; i < pPk->nKeyCol; i++){
+      pKey->aKeyAffinity[i] = pTable->aCol[pPk->aiColumn[i]].affinity;
+      pKey->azKeyCollation[i] = pPk->azColl[i];
+    }
+  } else{
+    pKey->nKeyColumns = 1;
+    pKey->aKeyAffinity[0] = SQLITE_AFF_INTEGER;
+    pKey->azKeyCollation[0] = "BINARY";
+  }
+  return 0;
+}
+
+int vectorIdxKeyDefsRender(const VectorIdxKey *pKey, const char *prefix, char *pBuf, int nBufSize) {
+  static const char * const azType[] = {
+    /* SQLITE_AFF_BLOB    */ " BLOB",
+    /* SQLITE_AFF_TEXT    */ " TEXT",
+    /* SQLITE_AFF_NUMERIC */ " NUMERIC",
+    /* SQLITE_AFF_INTEGER */ " INTEGER",
+    /* SQLITE_AFF_REAL    */ " REAL",
+    /* SQLITE_AFF_FLEXNUM */ " NUMERIC",
+  };
+  int i, size;
+  for(i = 0; i < pKey->nKeyColumns && nBufSize > 0; i++){
+    const char *collation = pKey->azKeyCollation[i];
+    if( sqlite3_strnicmp(collation, "BINARY", 6) == 0 ){
+      collation = "";
+    }
+    if( i == 0 ){
+      size = snprintf(pBuf, nBufSize, "%s %s %s", prefix, azType[pKey->aKeyAffinity[i] - SQLITE_AFF_BLOB], collation);
+    }else {
+      size = snprintf(pBuf, nBufSize, ",%s%d %s %s", prefix, i, azType[pKey->aKeyAffinity[i] - SQLITE_AFF_BLOB], collation);
+    }
+    if( size < 0 ){
+      return -1;
+    }
+    pBuf += size;
+    nBufSize -= size;
+  }
+  if( nBufSize <= 0 ){
+    return -1;
+  }
+  return 0;
+}
+
+int vectorIdxKeyNamesRender(int nKeyColumns, const char *prefix, char *pBuf, int nBufSize) {
+  int i, size;
+  for(i = 0; i < nKeyColumns && nBufSize > 0; i++){
+    if( i == 0 ){
+      size = snprintf(pBuf, nBufSize, "%s", prefix);
+    }else {
+      size = snprintf(pBuf, nBufSize, ",%s%d", prefix, i);
+    }
+    if( size < 0 ){
+      return -1;
+    }
+    pBuf += size;
+    nBufSize -= size;
+  }
+  if( nBufSize <= 0 ){
+    return -1;
+  }
+  return 0;
+}
+
+/**************************************************************************
+** VectorInRow utilities
+****************************************************************************/
+
+sqlite3_value* vectorInRowKey(const VectorInRow *pVectorInRow, int iKey) {
+  assert( 0 <= iKey && iKey < pVectorInRow->nKeys );
+  return pVectorInRow->pKeyValues + iKey;
+}
+
+i64 vectorInRowLegacyId(const VectorInRow *pVectorInRow) {
+  if( pVectorInRow->nKeys == 1 && sqlite3_value_type(pVectorInRow->pKeyValues + 0) == SQLITE_INTEGER ){
```

Comment:
`+ 0`?

Reply:
Yes, this is bit convoluted. But, `pVectorInRow->pKeyValues` is an array actually and here I wanted to highlight that we are accessing first element of this array.

I can write something like `&pVectorInRow->pKeyValues[0]` instead - but not sure if this will make more sense...

Reply:
[0] is probably a bit more idiomatic - I don't know